contracts/Gate.sol

```diff
@@ -38,9 +50,116 @@ contract Gate is ProtocolVersioned {
+
+    // stake requests
     mapping(address /*staker */ => StakeRequest) public stakeRequests;
 
-    function Gate() public { //TODO: complete as necessary
+    // bounty amount
+    uint256 private bounty;
+
+    // utility token UUID
+    bytes32 private uuid;
+
+    // OpenSTValue contract
+    OpenSTValue private openSTValue;
+
+    /*
+     *  Public functions
+     */
+    constructor(
+        address _workers,
+        uint256 _bounty,
+        bytes32 _uuid,
+        address _openSTValue)
+        public
+        ProtocolVersioned(_openSTValue)
+    {
+        require(_workers != address(0));
+        require(_openSTValue != address(0));
+        require(_uuid.length != 0);
+
+        // Should bounty with 0 amount be allowed ?
+        //require(_bounty > 0);
+
+        workers = _workers;
+        bounty = _bounty;
+        uuid = _uuid;
+        openSTValue = OpenSTValue(_openSTValue);
+    }
+
+    function requestStake(
+        uint256 _amount,
+        address _beneficiary)
+        external
+        returns ( bool /* success */)
+    {
+
+        require(_amount > 0);
+        require(_beneficiary != address(0));
+
+        // check if the stake request does not exists
+        require(stakeRequests[msg.sender].beneficiary == address(0));
+
+        require(openSTValue.valueToken().allowance(msg.sender, address(this)) >= _amount);
+        require(openSTValue.valueToken().transferFrom(msg.sender, address(this), _amount));
+
+        stakeRequests[msg.sender] = StakeRequest({
+            amount: _amount,
+            beneficiary: _beneficiary,
+            hashLock: 0,
+            unlockHeight: 0
+        });
 
+        emit StakeRequested(msg.sender, _amount, _beneficiary);
+
+        return true;
+    }
+
+    function revertStakeRequest()
+        external
+        returns (bool /* success */)
+    {
+        StakeRequest storage stakeRequest = stakeRequests[msg.sender];
+
+        // check if the stake request exists for the msg.sender
+        require(stakeRequest.beneficiary != address(0));
+
+        // check if the stake request was not accepted
+        require(stakeRequest.unlockHeight == 0);
+
+        require(openSTValue.valueToken().transfer(msg.sender, stakeRequest.amount));
+
+        delete stakeRequests[msg.sender];
+
+        emit StakeRequestReverted(msg.sender, stakeRequest.amount);
+
+        return true;
     }
+
+    function rejectStakeRequest(address _staker)
+        external
+        returns (bool /* success */)
+    {
+        // check if the caller is whitelisted worker
+        //require(workers.isWorker(msg.sender)); //TODO: revist this to add worker check
+
+        StakeRequest storage stakeRequest = stakeRequests[_staker];
+
+        // check if the stake request exists
+        require(stakeRequest.beneficiary != address(0));
+
+        // check if the stake request was not accepted
+        require(stakeRequest.unlockHeight == 0);
```

Comment:
similar correction for the definition of the state, let's use `hashLock != 0` rather than `unlockHeight`

Reply:
changed to ``` hashLock = 0x0 ``` instead of unlockHeight